doc/dev/osd_internals/multi_object_write_transaction.rst

```diff
@@ -0,0 +1,113 @@
+==============================
+Multi-Object Write Transaction
+==============================
+
+Multi-object write transaction enables a group of write operations
+on multi-objects complete atomically.
+
+Description of the Process
+--------------------------
+
+The high level process is,
+
+  1. Define the following class,
+
+     class *CEPH_RADOS_API MultiObjectWriteOperation*
+
+     {
+
+       std::map<std::string, ObjectWriteOperation*> slaves;
+
+       ObjectWriteOperation master;
+
+     };
+
+     based on this, client could send a group of ops on multi-objects
+     through *MOSDOp* to the primary osd in charge of the *master* object
+     which is explicitly specified by client, call this osd *MASTER*
+
+  #. *MASTER* recieves the multi-object write operation request, writes a
+     *LOCK* entry into *PGLOG*, creates a meta object, and records the *slaves*
+     information in it;
+
+  #. *MASTER* extracts operations by objects from *slaves*, enclose them in
+     *LOCK* requests, each request includes a group of operations on a single
+     slave object. *MASTER* sends *LOCK* requests to corresponding primary osds
+     individually, call those osds "SLAVE*;
+
+  #. *SLAVE* recieves *LOCK* request, confirms the validity of the operation,
+     then writes a *LOCK* entry into *PGLOG*, creates a meta object, and
+     records the transaction information in it, then sends ack to *MASTER*;
+
+  #. *MASTER* collects response, then does the operations on the *master*
+     object, besides, writes a *COMMIT* entry into *PGLOG*. *MASTER* sends
+     *COMMIT* request to *SLAVE*, then sends ack to client;
+
+  #. *SLAVE* receives *COMMIT*, performs the operations on the slave object,
+     besides, writes a *COMMIT* entry into *PGLOG*. *SLAVE* sends ack to
+     *MASTER*, deletes the meta object and writes a *UNLOCK* entry into
+     *PGLOG*;
```

Comment:
We could remove the operation of "write unlock entry", then the process is: commit and write commit entry in a transaction; ack sent to master; delete the meta object, so logically remove the UNLOCK state from the state machine. But in the current data structure design, the delete meta object operation could not be merged in the commit step, since one ctx could not touch two objects